build.assets/Dockerfile-multiarch

```diff
@@ -0,0 +1,98 @@
+# #############################################################################
+# This Dockerfile aims to be the single source of truth for linux buildboxes on
+# all supported architectures. This is still wildly aspirational, and 
+#   a) currently only supports ARM32 & ARM64, and
+#   b) is currently only used for ARM64 builds
+# #############################################################################
+
+FROM centos:7 AS base
+
+# Automatically supplied by the Docker buildkit
+ARG TARGETARCH
+
+# #############################################################################
+# Platform-specific customisation.
+# #############################################################################
+
+FROM base AS platform-setup-arm
+ARG GOLANG_ARCH=armv6l
+ARG YUM_UPDATE_FLAGS=""
+RUN echo "armhfp" > /etc/yum/vars/basearch && \
+    echo "armv7hl" > /etc/yum/vars/arch && \
+    echo "armv7hl-redhat-linux-gnu" > /etc/rpm/platform 
+
+FROM base AS platform-setup-arm64
+ARG GOLANG_ARCH=arm64
+
+# Installing the kerenel packages causes the update to hang on aarch64, so we
+# skip upgrading them.
+ARG YUM_UPDATE_FLAGS="--exclude kernel-*"
+RUN echo "aarch64-redhat-linux-gnu" > /etc/rpm/platform 
+
+# #############################################################################
+# The minimal buildbox contains the resources required to build a minimal 
+# version of Teleport, with no optional features (e.g. BPF)
+# #############################################################################
+
+FROM platform-setup-$TARGETARCH AS minimal-buildbox
```

Comment:
To use an ARG in a FROM line, the ARG needs to be declared before the first FROM line, according to https://docs.docker.com/engine/reference/builder/#scope:

> An ARG instruction goes out of scope at the end of the build stage where it was defined. To use an arg in multiple stages, each stage must include the ARG instruction.

and from https://docs.docker.com/engine/reference/builder/#understand-how-arg-and-from-interact

> FROM instructions support variables that are declared by any ARG instructions that occur before the first FROM.

There was one blog post I read that said you would inherit ARGs in multistage builds, but that seems contrary to the docs linked above.

This stage also uses `GOLANG_ARCH` which is also only declared in a previous stage.

Reply:
Also, tried declaring the `TARGETARCH` arg _before_ the first `FROM` and it didn't work: complained about not being able to find the `platform-setup-`, suggesting that `$TARGETARCH` was undefined.

Reply:
If you need an arg both before and after a `FROM` line, you need to redeclare the `ARG` with `ARG TARGETARCH` as described here: https://docs.docker.com/engine/reference/builder/#understand-how-arg-and-from-interact

Reply:
> `GOLANG_ARCH` is defined in both `platform-setup-*` stages. I've added some markers to help delineate the per-platform stages

Sure, but according to the docs, those ARGs go out of scope at the end of the build stage where they are defined (the bit quoted above). Perhaps it works if the implementation doesn't match the spec :shrug: , but you are using an ARG that is not defined in the build stage in which it is used, so should be out of scope according to the docs. Maybe I'm misreading the docs - if it all seems to work, then ok I guess.